### ISSUE TITLE
Annoying mac os fatal error on session.cleanup()

### ISSUE THREAD

**Issue by paulborgermans** [irods/python-irodsclient#657]
Hi

I am a bit surprised I did not find a previous issue for this, but cleanup() of an iRODSSession instance often results in a fatal error on Mac OS. The reason appears a particular intolerance on socket close handling in Mac OS (and also other BSD variants for that matter)

```
Traceback (most recent call last):
  File "/Users/paulborgermans/projects/playground-python/simple_test.py", line 14, in <module>
    with iRODSSession(irods_env_file=env_file, **ssl_settings) as session:
  File "/opt/anaconda3/lib/python3.12/site-packages/irods/session.py", line 169, in __exit__
    self.cleanup()
  File "/opt/anaconda3/lib/python3.12/site-packages/irods/session.py", line 203, in cleanup
    conn.disconnect()
  File "/opt/anaconda3/lib/python3.12/site-packages/irods/connection.py", line 330, in disconnect
    self.socket.shutdown(socket.SHUT_RDWR)
  File "/opt/anaconda3/lib/python3.12/ssl.py", line 1291, in shutdown
    super().shutdown(how)
OSError: [Errno 57] Socket is not connected
```

Would it be an option to put a try .. except block around self.socket.shutdown(socket.SHUT_RDWR) in connection.py?

I currently have to advise our users on Mac OS to put a try .. except block around a forced irods_session.cleanup() as the fatal error is of course also happening in a with block (practically unusable, unless itself wrapped in a try .. except block.

Thanks for your input
Paul

**Comment by korydraughn** [comment]
Q. What version of the PRC are you using?
Q. Does this always happen when `cleanup()` is invoked?
Q. Are there any other factors needed to trigger this other than use of a Mac?
Q. Can you post a small example which reproduces the issue?

**Comment by trel** [comment]
I think SSL being in use is important.

**Comment by paulborgermans** [comment]
Hi Kory, terrell,

It only happens on Mac OS, not Windows or Linux
I tested with PRC 2.0.1 and 2.2.0
Python versions 3.10 and 3.12

```
import os
import ssl
from irods.session import iRODSSession

env_file = os.path.expanduser('~/.irods/irods_environment.json')

ssl_context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH, cafile=None, capath=None, cadata=None)
ssl_settings = {'ssl_context': ssl_context}

session = iRODSSession(irods_env_file=env_file, **ssl_settings)

# Loop over many catalog items
for root, dirs, files in session.collections.get("/<zone>/home/<large collection with deeper nested structure>").walk():
    print('{root}-{dirs}-{files}'.format(root=root, dirs=dirs, files=files))

session.cleanup()
```

For smaller collection trees, it doe not always happen, for larger collections ( several 1000s of data objects and 4-6 levels deep), it always happens in my tests on Mac OS

**Comment by paulborgermans** [comment]
And I should have added: the walk itself has no errors, it is the session.cleanup() that throws a fatal error

**Comment by paulborgermans** [comment]
And to put the perspective right: for me it is either Python ssl module or Mac OS that has the bug, not PRC. I would love to see a hardened version of PRC that can cope with this without propagating the exception higher up in the food chain ;-)

**Comment by paulborgermans** [comment]
Here is an older report on a similar exception in another library, only happening on Mac OS:

https://github.com/kmonsoor/py-amqplib/issues/45

**Comment by trel** [comment]
looks... related...

https://bugs.python.org/issue4397 (ported to https://github.com/python/cpython/issues/48647)

good thing it was fixed in 2.6, 3.0, and 3.1 in Jan 2009.   :P

**Comment by trel** [comment]
I have reproduced the occasional error with your minimal example (tweaked below).
required SSL on the server.
running on Mac, PRC 2.2.0+ and Python 3.12.7...
only 17 data objects in the catalog
looped a few times... and it failed

```
import os
from irods.session import iRODSSession
env_file = os.path.expanduser("~/.irods/irods_environment.json")
ssl_settings = {}
with iRODSSession(irods_env_file=env_file, **ssl_settings) as session:
    for root, dirs, files in session.collections.get("/tempZone/home/rods").walk():
        print('{root}-{dirs}-{files}'.format(root=root, dirs=dirs, files=files))
```

**Comment by trel** [comment]
> Would it be an option to put a try .. except block around self.socket.shutdown(socket.SHUT_RDWR) in connection.py?

yes, looks like a viable solution is https://github.com/ethereum/trinity/pull/948/commits/b5e3d40fd4ba43464ae41a27ab1cbc9fe6d30dae

```diff
--- a/irods/connection.py
+++ b/irods/connection.py
@@ -325,7 +325,14 @@ class Connection:
                     self.socket = self.socket.unwrap()
                 except AttributeError:
                     pass
-                self.socket.shutdown(socket.SHUT_RDWR)
+                try:
+                    self.socket.shutdown(socket.SHUT_RDWR)
+                except OSError as e:
+                    import errno
+                    # On MacOS this can result in the following error:
+                    # OSError: [Errno 57] Socket is not connected
+                    if e.errno != errno.ENOTCONN:
+                        raise
                 self.socket.close()
         finally:
```

**Comment by trel** [comment]
proved the above fixes the issue - see https://github.com/irods/python-irodsclient/pull/658

**Comment by d-w-moore** [comment]
> proved the above fixes the issue - see #658

Nice!
